Mathlib/Computability/ContextFreeGrammar.lean

```diff
@@ -187,3 +190,73 @@ lemma Derives.append_right {v w : List (Symbol T g.NT)}
   | tail _ last ih => exact ih.trans_produces <| last.append_right p
 
 end ContextFreeGrammar
+
+/-- Context-free languages are defined by context-free grammars. -/
+def Language.IsContextFree (L : Language T) : Prop :=
+  ∃ g : ContextFreeGrammar.{uT} T, g.language = L
+
+section closure_reversal
+
+/-- Rules for a grammar for a reversed language. -/
+def ContextFreeRule.reverse {N : Type uN} (r : ContextFreeRule T N) : ContextFreeRule T N :=
+  ⟨r.input, r.output.reverse⟩
+
+/-- Grammar for a reversed language. -/
+def ContextFreeGrammar.reverse (g : ContextFreeGrammar T) : ContextFreeGrammar T :=
+  ⟨g.NT, g.initial, g.rules.map .reverse⟩
+
+lemma ContextFreeRule.reverse_reverse {N : Type uN} :
+    Function.Involutive (@ContextFreeRule.reverse T N) := by
+  intro x
+  simp [ContextFreeRule.reverse]
+
+lemma ContextFreeGrammar.reverse_reverse :
```

Comment:
The change to use `Function.Involutive` is good, but now these lemmas should be called `reverse_involutive`